drafts/multiformats.txt

```diff
@@ -0,0 +1,1008 @@
+   | 32         | BJV2WY5DJMJQXGZJANFZSAYLXMVZW63LFEEQFY3ZP           |
+   | 58         | zYAjKoNbau5KiqmHPmSxYCvn66dA1vLmwbt                 |
+   | 64         | MTXVsdGliYXNlIGlzIGF3ZXNvbWUhIFxvLw==               |
+   +------------+-----------------------------------------------------+
+
+3.2.  Multibase ABNF
+
+   ABNF
+
+   base-identifier = <ascii character>
+   multibase = base-identifier {base-encoded-data}
+
+4.  Multihash
+
+   Multihash is a format for differentiating outputs from various well-
+   established cryptographic hash functions while adressing output
+   length and encoding considerations.
+
+   Multihash values encode the hash function output prefixed by an id
+   identifying the hash function encoded as an unsigned variable length
+
+
+
+Snell                     Expires May 29, 2020                  [Page 4]
+
+Internet-Draft                Multiformats                 November 2019
+
+
+   integer, and the length of the hash function output encoded as an
+   unsigned variable length integer.  All valid Multihash function
+   identifiers are registered in the Codec Registry.
+
+   Multihash values are byte sequences that MAY be encoded in multiple
+   base-encodings including hex, base64, base32, or others.  When
+   representing a Multihash using a base-encoding, the Multibase format
+   SHOULD be used.
+
+4.1.  Example Multihashes
+
+   Given the sample text "multihash", the following example Multihash
+   values may be generated:
+
+   +--------+-----+----------------------------------------------------+
+   | Functi | ID  | Multihash (Multibase/Base16)                       |
+   | on     |     |                                                    |
+   +--------+-----+----------------------------------------------------+
+   | sha1   | 0x1 | f111488c2f11fb2ce392acb5b2986e640211c4690073e      |
+   |        | 1   |                                                    |
+   | sha2-2 | 0x1 | f12209cbc07c3f991725836a3aa2a581ca2029198aa420b9d9 |
+   | 56     | 2   | 9bc0e131d9f3e2cbe47                                |
+   +--------+-----+----------------------------------------------------+
+
+4.2.  Multihash ABNF
+
+   Multihash
+
+   hash-function-code = unsigned-varint
+   hash-digest-size = unsigned-varint
+   hash = {hash function output}
+
+   multihash = hash-function-code hash-digest-size hash
+
+5.  Security Considerations
+
+   There are no additional security considerations introduced by the use
+   of Multiformat data formats.  Note, however, that various hash
+   functions that may be used with multiformats may be vulnerable to
+   various known security issues and limitations (e.g. sha1) and thus
+   SHOULD be avoided.
+
+6.  IANA Considerations
+
+
+
+
+
+
+
+
+Snell                     Expires May 29, 2020                  [Page 5]
+
+Internet-Draft                Multiformats                 November 2019
+
+
+6.1.  The Codec Registry
```

Comment:
I keep on tripping up on this "codec" / multicodec / multihash soup, with multibase standing on the sidelines and I don't think I'll be alone on this.

I'd love to hear from someone who's been working on these things for a lot longer than I how your mental model works for how they piece together. I think we need to do some work bridging between that and how new folks will be building their mental models.

If I were to try and explain it to someone who hasn't seen this stuff before, although I don't think this clears it up enough:

 * "multiformats" is an umbrella term for this whole concept of self-describing data that involves prefixing binary strings with (mainly) varint codes
 * "multicodec" is a term that primarily refers to the table of codes and is encountered where you need to identify the type of a thing - CID being its primary direct user where the code determines the type of thing being linked to. The way I see it used makes me think that "multicodec" is a poor name for it and it's really just a registry.
 * a "format" is how you use the codes from the "multicodec" registry; so we end up with "multihash" and "multibase" as the two most heavily used ones forms, and they are distinct. I suppose CID might be another "format" that munges them together in a special way. IPLD could be another, but it only interacts via CID. Although neither of these have primary entries in the "multicodec" registry.
 * the "multicodec" table contains codes that are used by "multihash" but not "multibase", it has its own table, maybe for historical reasons?
 * the "multicodec" table contains codes for "multihash", "multibase" and even "multicodec", but I don't think I've seen those used in the wild because when you consume these things via something like CID you're dealing with predictable positioning. Although maybe you could link to a thing with a CID that's actually a multihash itself?

Reply:
> * "multicodec" is a term that primarily refers to the table of codes and is encountered where you need to identify the type of a thing - CID being its primary direct user where the code determines the type of thing being linked to. The way I see it used makes me think that "multicodec" is a poor name for it and it's really just a registry.

"multicodec" used to describe something that is prefix + data. Though this wasn't how it is actually used these days. Hence I adapted to "multicodec" spec to describe what it actually does, it is just a registry.

> * the "multicodec" table contains codes that are used by "multihash" but not "multibase", it has its own table, maybe for historical reasons?

That is also a more recent development. They used to be in the same table, but that was really confusing (and also lead to wrong implementations), hence they were splitted into two. The codec in "multicodec" describe byte values, the codecs in "multibase" describe characters, which depend on the string encoding, hence are not always the same bytes.